### DISCUSSION TITLE
Low performance of a single thread.

### DISCUSSION THREAD

**Discussion by stm32repo** [dragonflydb/dragonfly#1688] [Q&A]
I launched dragonflydb and Redis on a machine with a database of 100 million records. Then, I conducted performance tests using the redis-benchmark utility and obtained significantly worse results from Dragonfly compared to Redis when running a single thread.

Port 6379 - Redis
Port 16379 - Dragonflydb
```
redis-benchmark -p 6379 -t get -c 1 -n 1000000 --dbnum 1 -r 20 -P 1
Summary:
  throughput summary: 32072.87 requests per second
  latency summary (msec):
          avg       min       p50       p95       p99       max
        0.027     0.008     0.031     0.039     0.047     0.543
		
redis-benchmark -p 6379 -t get -c 1 -n 1000000 --dbnum 1 -r 20 -P 100
Summary:
  throughput summary: 1070663.75 requests per second
  latency summary (msec):
          avg       min       p50       p95       p99       max
        0.080     0.048     0.079     0.119     0.127     0.223
		
redis-benchmark -p 6379 -t get -c 20 -n 1000000 --dbnum 1 -r 20 -P 1 --threads 20
Summary:
  throughput summary: 107968.04 requests per second
  latency summary (msec):
          avg       min       p50       p95       p99       max
        0.177     0.024     0.167     0.263     0.327    13.399
				
redis-benchmark -p 16379 -t get -c 1 -n 1000000 --dbnum 1 -r 20 -P 1
Summary:
  throughput summary: 18182.81 requests per second
  latency summary (msec):
          avg       min       p50       p95       p99       max
        0.052     0.016     0.055     0.071     0.079     1.535
		
redis-benchmark -p 16379 -t get -c 1 -n 1000000 --dbnum 1 -r 20 -P 100
Summary:
  throughput summary: 50027.52 requests per second
  latency summary (msec):
          avg       min       p50       p95       p99       max
        1.980     1.376     1.967     2.287     2.479     3.343
		
redis-benchmark -p 16379 -t get -c 20 -n 1000000 --dbnum 1 -r 20 -P 1 --threads 20
Summary:
  throughput summary: 249252.23 requests per second
  latency summary (msec):
          avg       min       p50       p95       p99       max
        0.074     0.008     0.071     0.127     0.175    25.279
```

**Comment by romange** [comment]
not a bug - but a design choice. 

we had many questions like these before. @Niennienzz  we should probably add a documentation page on how to benchmark Dragonfly so we could just reply with a single link. 

TLDR: if you intend to run your memory store locally or using a single thread - please continue using Redis. It has lower latency for a single connection. Also, there are not many performance advantages when running Dragonfly in a single thread. So if this is your use-case - then do not use Dragonfly.

Every implementation has its own best practices. 
Dragonfly can not parallelize efficiently workloads when serving a single connection because it needs to respect the order for the incoming requests on that connection. Dragonfly shines when there are multiple client connections. Dozens, hundreds - the more the merrier. Dragonfly also has limits internally of how much pipelining it allows. AFAIK, the factor is 10, meaning pipelining more than the limit won't produce more throughput. The reason for this is that pipelining adds more latency and more memory pressure and again, the best practice with Dragonfly is to parallelize using multiple connections. 

regarding load test methodology:
1. We perform benchmarking on separate machines for both the load test and the server. Dragonfly uses all the cpus by default, so it contends with the load test program if running on the same machine. 
2. We usually use `memtier_benchmark` and not `redis-benchmark` which is less efficient.
